src/primitives/hrpstring.rs

```diff
@@ -0,0 +1,447 @@
+// SPDX-License-Identifier: MIT
+
+//! String encoding/decoding of HRP strings as specified by [BIP-173] (bech32) and [BIP-350] (bech32m).
+//!
+//! HRP string format: `<hrp> 1 <witness_version> <payload> <checksum>`
+//!
+//! * **hrp**: Human Readable Part.
+//! * **witness_version**: 0 ('q) for bech32 1 ('p') for bech32m.
+//! * **payload**: GF(32) bech32 encoded characters.
+//! * **checksum**: BCH code checksum.
+//!
+//! [BIP-173]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki>
+//! [BIP-350]: <https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki>
+
+use core::convert::{Infallible, TryFrom};
+use core::{fmt, iter, slice, str};
+
+use crate::primitives::checksum::{self, Checksum};
+use crate::primitives::gf32::{self, Fe32};
+use crate::primitives::hrp::{self, Hrp};
+use crate::primitives::iter::{Fe32IterExt, Fe32ToByteIter};
+use crate::write_err;
+
+/// Separator between the hrp and payload (as defined by BIP-173).
+const SEP: char = '1';
+
+/// An HRP string that has been parsed and had the checksum validated.
+///
+/// Pre-parsing an HRP string only checks validity of the characters, it does not validate the checksum in any way - to validate convert to [`Parsed`].
+#[derive(Debug)]
+pub struct Parsed<'s> {
```

Comment:
In e2527dc19c3e8e9a4cd678a955fc5b161e175c87:

Here is a proposal:

* We make `Parsed` generic over the `Ck` by adding a phantomdata. This makes the API slightly less flexible but I think is worth it in clarity.
* We add an internal `checksum` field which stores the ASCII values of the checksum.
* Then `new` and `new_with_witness_version` would take `Ck` implicitly from the type that is being constructed.
* We make `validate_checksum` public and have it **re-validate the checksum** and potentially turn the `Parsed<Ck>` into a `Parsed<OtherCk>`.
* We replace `parse_unvalidated` with special constructors for `Parsed<NoChecksum>` which skip checksum validation entirely, and make **these** public as well. (We'd add a unit test that if you used `new`/`new_with_witness_version` with `NoChecksum`, that you got the same result as these.)
* We clearly document that `new` and `new_with_witness_version` are the only constructors that norma users need to worry about.

Then in src/lib.rs, to deal with bech32 vs bech32m, we can take a `Parsed::new_with_witness_version_no_checksum(s)?` and then alternately call `revalidate_checksum::<Bech32>` or `revalidate_checksum::<Bech32m>` and save all the unnecessary iterations.

Reply:
I did not get to processing this suggestion today - will come back to it.

Reply:
I had a brief play with this. Using checksum generic on `Parsed` borks the current API in `libr.s` by making it impossible, for example, to return `Parsed` from `decode`. It also defeats the purpose of having the `Variant` type (which I do think will go away eventually but I don't think we should do that yet).

Can we leave this stuff until we start working out what the final crate API looks like? I"ll save a link to the stuff above so it doesn't get lost.

Reply:
I'd really like to figure out how to use the API without doing twice as much work to parse segwit addresses as we used to.

Reply:
Ok, we can punt on this until a later stage when we get rid of `Variant` in place of using the `Ck` parameter.